File: ios/PlaygamaGames/Models/Game.swift
// Game.swift
// Data models for games from the Playgama catalog

import Foundation

// MARK: - Root JSON Structure

/// Root structure of the JSON game catalog file
struct GameCatalog: Codable {
    let segments: [GameSegment]
}

// MARK: - Segment (Game Category)

/// A segment is a category or collection of games
struct GameSegment: Codable {
    let title: String
    let count: Int
    let hits: [Game]
}

// MARK: - Game Model

/// Main game model from the Playgama catalog
struct Game: Codable, Identifiable {
    
    // MARK: Identification
    let id: String
    let slug: String
    let title: String
    
    // MARK: Description
    let description: String
    let howToPlayText: String?
    
    // MARK: URLs
    
    /// Main URL to load the game in WebView
    let gameURL: String
    let playgamaGameUrl: String
    
    // MARK: Media
    let images: [String]
    let videos: [GameVideo]?
    
    // MARK: Categories
    let genres: [String]
    let tags: [String]?
    
    // MARK: Compatibility
    let mobileReady: [String]?
    let supportedLanguages: [String]?
    let screenOrientation: ScreenOrientation?
    
    // MARK: Additional Info
    let gender: [String]?
    let inGamePurchases: String?
    let embed: String?
}

// MARK: - Supporting Structures

struct GameVideo: Codable {
    let playgama_id: String?
    let external_url: String?
    let type: String?
}

struct ScreenOrientation: Codable {
    let horizontal: Bool
    let vertical: Bool
}

// MARK: - Convenience Extensions

extension Game {
    
    var thumbnailURL: URL? {
        guard let firstImage = images.first else { return nil }
        return URL(string: firstImage)
    }
    
    var gamePlayURL: URL? {
        return URL(string: gameURL)
    }
    
    var supportsIOS: Bool {
        return mobileReady?.contains("For IOS") ?? false
    }
    
    var supportsAndroid: Bool {
        return mobileReady?.contains("For Android") ?? false
    }
    
    var preferredOrientation: String {
        guard let orientation = screenOrientation else { return "any" }
        if orientation.horizontal && !orientation.vertical {
            return "landscape"
        } else if orientation.vertical && !orientation.horizontal {
            return "portrait"
        }
        return "any"
    }
    
    var genresDisplayText: String {
        let displayGenres = genres.prefix(3)
        return displayGenres.joined(separator: " • ")
    }
}

File: android/app/src/main/java/com/playgama/games/models/Game.kt
package com.playgama.games.models

/**
 * Data models for games from the Playgama catalog.
 * 
 * The JSON catalog contains "segments" - game categories.
 * Each segment contains a list of games (hits).
 */

import com.google.gson.annotations.SerializedName

// ============================================================
// ROOT STRUCTURE
// ============================================================

/**
 * Root structure of the JSON catalog.
 * Contains a list of segments (categories).
 */
data class GameCatalog(
    val segments: List<GameSegment>
)

/**
 * A segment is a category or collection of games.
 * Example: "Popular games", "New", "Action", etc.
 */
data class GameSegment(
    val title: String,
    val count: Int,
    val hits: List<Game>
)

// ============================================================
// MAIN GAME MODEL
// ============================================================

/**
 * Main game model from the Playgama catalog.
 * 
 * Key fields:
 * - [gameURL] - main URL to load the game in WebView
 * - [images] - list of cover image URLs
 * - [genres] - list of game genres
 * - [mobileReady] - list of supported platforms
 */
data class Game(
    // Identification
    val id: String,
    val slug: String,
    val title: String,
    
    // Description
    val description: String,
    
    @SerializedName("howToPlayText")
    val howToPlayText: String? = null,
    
    // URLs
    /** Main URL to load the game in WebView */
    val gameURL: String,
    
    val playgamaGameUrl: String,
    
    // Media
    /** List of cover image URLs */
    val images: List<String>,
    
    val videos: List<GameVideo>? = null,
    
    // Categories
    /** List of game genres (e.g., "action", "puzzle", "arcade") */
    val genres: List<String>,
    
    val tags: List<String>? = null,
    
    // Compatibility
    /** Supported platforms (e.g., "For Android", "For IOS", "For Desktop") */
    val mobileReady: List<String>? = null,
    
    val supportedLanguages: List<String>? = null,
    
    /** Supported screen orientations */
    val screenOrientation: ScreenOrientation? = null,
    
    // Additional info
    val gender: List<String>? = null,
    val inGamePurchases: String? = null,
    val embed: String? = null
) {
    // ========================================================
    // CONVENIENCE PROPERTIES
    // ========================================================
    
    /** Returns the first image URL or null */
    val thumbnailUrl: String?
        get() = images.firstOrNull()
    
    /** Returns true if the game supports Android */
    val supportsAndroid: Boolean
        get() = mobileReady?.contains("For Android") == true
    
    /** Returns true if the game supports iOS */
    val supportsIOS: Boolean
        get() = mobileReady?.contains("For IOS") == true
    
    /** Returns the preferred screen orientation */
    val preferredOrientation: String
        get() {
            val orientation = screenOrientation ?: return "any"
            return when {
                orientation.horizontal && !orientation.vertical -> "landscape"
                orientation.vertical && !orientation.horizontal -> "portrait"
                else -> "any"
            }
        }
    
    /** Returns formatted genres for display (first 3) */
    val genresDisplayText: String
        get() = genres.take(3).joinToString(" • ")
}

// ============================================================
// SUPPORTING STRUCTURES
// ============================================================

/**
 * Video information for the game.
 */
data class GameVideo(
    @SerializedName("playgama_id")
    val playgamaId: String? = null,
    
    @SerializedName("external_url")
    val externalUrl: String? = null,
    
    val type: String? = null
)

/**
 * Supported screen orientations for the game.
 */
data class ScreenOrientation(
    val horizontal: Boolean,
    val vertical: Boolean
)
